lambdajs/src/SharedAbstract.hs

```diff
@@ -0,0 +1,262 @@
+import qualified Prelude
+import           SharedInterpreter
+import           Syntax
+
+import           Data.Hashable
+import           Data.Identifiable
+import           Data.Set
+
+import           Data.Abstract.Environment
+import           Data.Abstract.HandleError
+import           Data.Abstract.Store
+import           Data.List                                       (find)
+import           Data.Order
+
+import           Control.Arrow.Transformer.Abstract.Environment
+import           Control.Arrow.Transformer.Abstract.HandleExcept
+import           Control.Arrow.Transformer.Abstract.Store
+import           Control.Arrow.Transformer.State
+import           Control.Arrow.Utils                             (map, pi1, pi2)
+
+import           Control.Arrow
+import           Control.Arrow.Environment
+import           Control.Arrow.Except
+import           Control.Arrow.Fail
+import           Control.Arrow.Reader
+import           Control.Arrow.State
+import           Control.Arrow.Store
+import           Control.Category
+
+
+instance PreOrd Type where
+    TLambda ids1 _ e1 ⊑ TLambda ids2 _ e2 = ids1 == ids2 && e1 ⊑ e2
+    TObject fields1 ⊑ TObject fields2 = all (\((f1, t1), (f2, t2)) -> f1 == f2 && t1 ⊑ t2) (zip fields1 fields2)
+    TRef l1 ⊑ TRef l2 = Data.Set.isSubsetOf l1 l2
+    TThrown t1 ⊑ TThrown t2 = Data.Set.isSubsetOf t1 t2
+    TBreak l1 t1 ⊑ TBreak l2 t2 = Data.Set.isSubsetOf t1 t2 && l1 == l2
+    a ⊑ b | a == b = True
+    _ ⊑ _ = False
+
+
+
+newtype TypeArr x y = TypeArr
+    (Except
+        String
+        (Environment Ident Type'
+            (StoreArrow Location Type'
+                (State Location (->)))) x y)
+
+deriving instance ArrowFail String TypeArr
+deriving instance ArrowEnv Ident Type' (Env Ident Type') TypeArr
+deriving instance Category TypeArr
+deriving instance Arrow TypeArr
+deriving instance ArrowChoice TypeArr
+deriving instance ArrowRead Location Type' x Type' TypeArr
+deriving instance ArrowWrite Location Type' TypeArr
+deriving instance ArrowState Location TypeArr
+
+runType :: TypeArr x y -> [(Ident, Type)] -> [(Location, Type)] -> x -> (Location, (Store Location Type', Error String y))
+runType (TypeArr f) env env2 x = runState (runStore (runEnvironment (runExcept f))) (Location 0, (Data.Abstract.Store.fromList env2', (Data.Abstract.Environment.fromList env', x)))
+        where env' = Prelude.map (\(x, y) -> (x, Data.Set.fromList [y])) env
+              env2' = Prelude.map (\(x, y) -> (x, Data.Set.fromList [y])) env2
+
+runAbstract :: [(Ident, Type)] -> [(Location, Type)] -> Expr -> (Store Location Type', Error String Type')
+runAbstract env st exp = case runType eval env st exp of
+    (l, (st, Fail e))      -> (st, Fail e)
+    (l, (st, Success res)) -> (st, Success res)
+
+typeEvalBinOp_ :: (ArrowFail String c, ArrowChoice c) => c (Op, Type, Type) Type
+typeEvalBinOp_ = proc (op, v1, v2) -> case (op, v1, v2) of
+    -- number operators
```

Comment:
Is it possible to implement this as a pure pattern match (`arr x -> case x of ...`)? This would be more performant.

Reply:
Fixed(?), when the op doesn't match it fails, but this is not possible now. I kept the change though.